cores/nRF5/HardwarePWM.cpp

```diff
@@ -53,9 +53,120 @@ HardwarePWM* HwPWMx[] =
 #endif
 };
 
-HardwarePWM::HardwarePWM(NRF_PWM_Type* pwm)
+#if CFG_DEBUG
+bool can_stringify_token(uintptr_t token)
+{
+  uint8_t * t = (uint8_t *)&token;
+  for (size_t i = 0; i < sizeof(uintptr_t); ++i, ++t)
+  {
+    uint8_t x = *t;
+    if ((x < 0x20) || (x > 0x7E)) return false;
+  }
+  return true;
+}
+
+void HardwarePWM::DebugOutput(Stream& logger)
+{
+  const size_t count = arrcount(HwPWMx);
+  logger.printf("HwPWM Debug:");
+  for (size_t i = 0; i < count; i++) {
+    HardwarePWM const * pwm = HwPWMx[i];
+    uintptr_t token = pwm->_owner_token;
+    logger.printf(" || %d:", i);
+    if (can_stringify_token(token)) {
+      uint8_t * t = (uint8_t*)(&token);
+      static_assert(sizeof(uintptr_t) == 4);
+      logger.printf("   \"%c%c%c%c\"", t[0], t[1], t[2], t[3] );
+    } else {
+      static_assert(sizeof(uintptr_t) == 4);
+      logger.printf(" %08x", token);
+    }
+    for (size_t j = 0; j < MAX_CHANNELS; j++) {
+      uint32_t r = pwm->_pwm->PSEL.OUT[j]; // only read it once
+      if ( (r & PWM_PSEL_OUT_CONNECT_Msk) != (PWM_PSEL_OUT_CONNECT_Disconnected << PWM_PSEL_OUT_CONNECT_Pos) ) {
+        logger.printf(" %02x", r & 0x1F);
+      } else {
+        logger.printf(" xx");
+      }
+    }
+  }
+  logger.printf("\n");
+}
+#endif // CFG_DEBUG
+
+// returns true ONLY when (1) no PWM channel has a pin, and (2) the owner token is nullptr
+bool HardwarePWM::takeOwnership(uintptr_t token)
+{
+  if (token == 0) {
+    LOG_LV1("HwPWM", "zero / nullptr is not a valid ownership token (attempted use in takeOwnership)");
+    return false; // cannot take ownership with nullptr
+  }
+  if (token == this->_owner_token) {
+    LOG_LV1("HwPWM", "failing to acquire ownership because already owned by requesting token (cannot take ownership twice)");
+  }
+  if (this->_owner_token != 0) {
+    LOG_LV3("HwPWM", "failing to acquire ownership because already owned by other token");
+    return false;
+  }
+  if (this->usedChannelCount() != 0) {
+    LOG_LV3("HwPWM", "failing to acquire ownership because at least one channel connected");
+    return false;
+  }
+  if (this->enabled()) {
+    LOG_LV3("HwPWM", "failing to acquire ownership because peripheral is already enabled");
+    return false;
+  }
+  // TODO: warn, but do not fail, if taking ownership with IRQs already enabled
+  // NVIC_GetActive
+
+  // use gcc built-in intrinsic to ensure atomicity
+  // See https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html
+  return __sync_bool_compare_and_swap(&(this->_owner_token), 0, token);
```

Comment:
Can you help me understand why you think this is legacy?

IIUC, this is GCC's built-in intrinsic to a `compare-and-swap` function, using hardware-specific functionality, and that GCC ensures the most performant option that works for the platform is chosen.

Also .... you cannot use a mutex from an ISR.  The above intrinsic works anytime.

Reply:
@henrygab don't use those builtin for mutex, we should use the freeRTOS one, it support giving/taking mutex from ISR https://www.freertos.org/a00124.html

Reply:
You are requiring the use of a semaphore, rather than CSWAP.  This will needs deeper review.  I am very unlikely to have an answer in the next two weeks.  Areas that would need deep consideration include:

1. if recursive semaphore acquisition will need to be supported
2. performance of CSWAP vs. the final semaphore type
3. correctness of any new design

In the alternative, at any time prior to this deeper review, it seems the above is correct code, and thus could be accepted.

If this is **_not_** correct code, please help me understand any situation where it would not result in correct execution.

Reply:
tone() has been rewritten to fully avoid the need for ISR.  will update branch soon.

Reply:
Removed the "legacy" function, even though it was clearly **_not_** targeted to Intel CPUs.